api/lib/opentelemetry.rb

```diff
@@ -24,7 +24,18 @@ module OpenTelemetry
 
   attr_writer :tracer_provider, :meter_provider, :baggage
 
-  attr_accessor :logger
+  DEFAULT_LOG_LEVEL = Logger::INFO
+  private_constant :DEFAULT_LOG_LEVEL
+
+  def logger
+    return @@logger if defined?(@@logger)
+
+    self.logger = Logger.new(STDOUT, level: ENV['OTEL_LOG_LEVEL'] || DEFAULT_LOG_LEVEL)
+  end
+
+  def logger=(logger)
+    class_variable_set :@@logger, logger
+  end
```

Comment:
The `DEFAULT_LOG_LEVEL` constant seems unnecessary. Do we need the `logger` and `logger=` accessors instead of the `attr_accessor :logger` that we had previously?

Reply:
If implemented as an `attr_accessor` at module/class level (as originally written), tests were somehow managing to access the `OpenTelemetry.logger` property before it was set to a valid Logger instance.  

That is, the `self.logger = Logger.new(...)` line wasn't executing soon enough even though it was obviously part of the module's construct.  

So I moved it to be lazily evaluated and set on a class variable and it resolved the test failures. 

```
  1) Error:
OpenTelemetry::SDK::Trace::Span::#finish#test_0005_does not allow ending more than once:
NoMethodError: private method `warn' called for nil:NilClass
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/span.rb:193:in `block in finish'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/span.rb:191:in `synchronize'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/span.rb:191:in `finish'
    ~/opentelemetry-ruby/sdk/test/opentelemetry/sdk/trace/span_test.rb:272:in `block (3 levels) in <top (required)>'

  2) Error:
OpenTelemetry::SDK::Trace::TracerProvider::#shutdown#test_0003_only notifies the span processor once:
NoMethodError: private method `warn' called for nil:NilClass
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/tracer_provider.rb:55:in `block in shutdown'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/tracer_provider.rb:53:in `synchronize'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/tracer_provider.rb:53:in `shutdown'
    ~/opentelemetry-ruby/sdk/test/opentelemetry/sdk/trace/tracer_provider_test.rb:44:in `block (3 levels) in <top (required)>'

  3) Error:
OpenTelemetry::SDK::Configurator::#configure::instrumentation installation#test_0001_installs single instrumentation:
NoMethodError: private method `warn' called for nil:NilClass
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:82:in `rescue in install_instrumentation'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:75:in `install_instrumentation'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:50:in `block (2 levels) in install'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:46:in `each'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:46:in `block in install'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:45:in `synchronize'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:45:in `install'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/configurator.rb:114:in `install_instrumentation'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/configurator.rb:97:in `configure'
    ~/opentelemetry-ruby/sdk/test/opentelemetry/sdk/configurator_test.rb:238:in `block (4 levels) in <top (required)>'

  4) Error:
OpenTelemetry::SDK::Configurator::#configure::instrumentation installation#test_0002_installs all:
NoMethodError: private method `warn' called for nil:NilClass
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:82:in `rescue in install_instrumentation'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:75:in `install_instrumentation'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:63:in `block (2 levels) in install_all'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:62:in `each'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:62:in `block in install_all'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:61:in `synchronize'
    ~/opentelemetry-ruby/api/lib/opentelemetry/instrumentation/registry.rb:61:in `install_all'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/configurator.rb:116:in `install_instrumentation'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/configurator.rb:97:in `configure'
    ~/opentelemetry-ruby/sdk/test/opentelemetry/sdk/configurator_test.rb:249:in `block (4 levels) in <top (required)>'

  5) Error:
OpenTelemetry::SDK::Trace::Span::#add_event#test_0009_does not add an event if span is ended:
NoMethodError: private method `warn' called for nil:NilClass
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/span.rb:104:in `block in add_event'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/span.rb:102:in `synchronize'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/span.rb:102:in `add_event'
    ~/opentelemetry-ruby/sdk/test/opentelemetry/sdk/trace/span_test.rb:147:in `block (3 levels) in <top (required)>'

  6) Error:
OpenTelemetry::SDK::Trace::Span::#status=#test_0002_does not set the status if span is ended:
NoMethodError: private method `warn' called for nil:NilClass
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/span.rb:144:in `block in status='
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/span.rb:142:in `synchronize'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/span.rb:142:in `status='
    ~/opentelemetry-ruby/sdk/test/opentelemetry/sdk/trace/span_test.rb:226:in `block (3 levels) in <top (required)>'

  7) Error:
OpenTelemetry::SDK::Trace::Export::BatchSpanProcessor::export retry#test_0001_should not retry on FAILURE exports:
NoMethodError: undefined method `error' for nil:NilClass
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb:139:in `report_result'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb:129:in `export_batch'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb:123:in `block in work'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb:114:in `loop'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb:114:in `work'
    ~/opentelemetry-ruby/sdk/lib/opentelemetry/sdk/trace/export/batch_span_processor.rb:59:in `block in initialize'

  8) Failure:
OpenTelemetry::SDK::Configurator::#logger#test_0001_returns a logger instance [~/opentelemetry-ruby/sdk/test/opentelemetry/sdk/configurator_test.rb:14]:
Expected nil to be an instance of Logger, not NilClass.
```

It seemed to me that the only reason it appeared to work previously without any errors is because of line 266 in sdk/test/opentelemetry/sdk/configurator_test.rb where we're "resetting globals" and call `OpenTelemetry.logger = Logger.new(STDOUT)` (which also was the reason we couldn't successfully redirect log output to "/dev/null"

Actually, now that I look at that reset_globals more closely, I see why my original attempt to follow the prevailing implementation pattern was failing to work.  The method is resetting all instance variables of the `OpenTelemetry` module back to nil.  So switching the implementation to a class level variable was successfully bypassing the global_resets!

I will look at resolving how globals get reset between unit tests to prevent undoing intended default behaviors and solving this should allow me to revert to the much simpler class-level accessor implementation that was there originally and have clean tests as well.

Reply:
Changed logger implementation to be congruent with all the other accessors presently implemented.

The key was removing the reset_globals in the unit test, which was overly aggressive.  only one test depended on an ivar reset and that was the instrumentation_registry property.  This is now specifically reset in the before block for its tests.